src/analysis/history.hpp

```diff
@@ -58,21 +61,71 @@ Real ReduceOneVar(MeshData<Real> *md, const std::string &varname, int idx = 0) {
   // We choose to apply volume weighting when using the sum reduction.
   // This assumes that for "Sum" variables, the variable is densitized, meaning
   // it already contains a factor of the measure: sqrt(det(gamma))
+  Real result = 0.0;
+  Reducer_t reducer(result);
+
   const bool volume_weighting =
-      std::is_same<Reducer_t, Kokkos::Sum<Real, HostExecSpace>>::value;
+      std::is_same<Reducer_t, Kokkos::Sum<Real, HostExecSpace>>::value ||
+      std::is_same<Reducer_t, Kokkos::Sum<Real, DevExecSpace>>::value ||
+      std::is_same<Reducer_t, Kokkos::Sum<Real, Kokkos::DefaultExecutionSpace>>::value;
 
+  parthenon::par_reduce(
+      parthenon::LoopPatternMDRange(), "Phoebus History for " + varname, DevExecSpace(),
+      0, pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &lresult) {
+        // join is a Kokkos construct
+        // that automatically does the
+        // reduction operation locally
+        const auto &coords = pack.GetCoords(b);
+        const Real vol = volume_weighting ? coords.CellVolume(k, j, i) : 1.0;
+        reducer.join(lresult, pack(b, ivar.first + idx, k, j, i) * vol);
+      },
+      reducer);
+  return result;
+}
+
+template <typename Reducer_t>
+Real ReduceInGain(MeshData<Real> *md, const std::string &varname, int idx = 0) {
+  const auto ib = md->GetBoundsI(IndexDomain::interior);
+  const auto jb = md->GetBoundsJ(IndexDomain::interior);
+  const auto kb = md->GetBoundsK(IndexDomain::interior);
+
+  PackIndexMap imap;
+  namespace iv = internal_variables;
+  std::vector<std::string> vars = {varname, iv::GcovHeat::name(), iv::GcovCool::name()};
+  const auto pack = md->PackVariables(vars, imap);
+  const auto ivar = imap[varname];
+  const auto iheat = imap[iv::GcovHeat::name()].first;
+  const auto icool = imap[iv::GcovCool::name()].first;
+  auto *pmb = md->GetParentPointer();
```

Comment:
switch to sparse packs. Also just pull both heat and cool out here, rather than passing them in is as string arguments above.

Reply:
done